app/models/access_token.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+class AccessToken < ApplicationRecord
+  belongs_to :user
+
+  # Bearerトークンとしての利用を想定しているため、フォーマットはBearerトークンのシンタックスを満たす必要がある。
+  # https://datatracker.ietf.org/doc/html/rfc6750#section-2.1
+  #
+  # 上記では、トークンの文字列長は1以上であればよいことになっているが、短すぎると総当たりで突破されないとも
+  # 限らないので 保険として、 `=` によるパディング以外の部分に関して、ある程度の長さを要求する。
+  validates :token, format: { with: %r{\A[\w\-.~+/]{16,}=*\Z} }, uniqueness: true
+  validates :scope, format: { with: /\A(READ|WRITE)( (READ|WRITE)*)\Z/ }
+  validates :expires_in, numericality: { greater_than_or_equal_to: 0 }
+
+  # 新規レコードの場合かつ属性が引数で明示的に設定されていない場合はデフォルト値を設定する。
+  def initialize(attributes = nil)
```

Comment:
デフォルト引数を Hash にしたらいろいろはかどるのでは?

```suggestion
  def initialize(attributes = {})
```

Reply:
親と思われる関数ではデフォルト値がnilだったので、そこに合わせました。

https://api.rubyonrails.org/classes/ActiveRecord/Core.html#method-c-new

Reply:
多分それなりに新しい Ruby 処理系使って今の開発って進めてもらってますよね？

↓の書き方ってどこかで deprecated になってなかったかしら。
```ruby
def initialize(attributes = {})
```

```ruby
def initialize(**attributes)
```
のほうが最近っぽい気が

Reply:
この辺りですかね
- https://docs.rubocop.org/rubocop/cops_style.html#styleoptionhash
- https://rubystyle.guide/#keyword-arguments-vs-option-hashes

Reply:
つい古い Ruby 処理系をサポートしたくなる癖が出たけど、いまなら `**attributes` 一択ですね…

※ ツールとかフレームワークを作る側なので、ひろく対応したくなっちゃうマンです

Reply:
↑のような.initializeのオーバーライド時のデフォルト引数の変更を試みたのですが、現時点では変更は事実上不可能というオチでした。

というのも、ActiveRecord::Baseを継承しているクラスが.newを呼ぶと、以下のメソッドを呼ぶことになります。そして、このメソッドのデフォルト引数がnilに設定されており、かつその引数が最終的に子クラスの.initializeに渡されると思われます。本当は、ここについてはまだコードを追いきれていないですが、実験やデバッガで確認した範囲だと、こういう理解でよさそうでした。つまり、.newを引数省略で呼んだとしても、呼び出し順の上位のメソッドのデフォルト引数であるnilが渡されることになるので、↑のように.initializeメソッドのオーバーライド時にデフォルト引数を設定しても意味がないということでした。

https://github.com/rails/rails/blob/e446ef82c8174a4bd6c98f670912bb395b3091d6/activerecord/lib/active_record/inheritance.rb#L56

なので、もし.initializeメソッドをオーバーライドする必要があるのなら、デフォルト引数を設定してもほとんど意味はないが、引数省略時にnilが渡されるという挙動を暗示できるため、オーバーライド時でもnilを設定しておくといいかもしれない、ということでした。

あと、**演算子を利用したメソッド定義に関しては、上記のメソッドが旧来のスタイルを利用して定義されている都合上、キーワード引数ではなく一つのハッシュとして引数を渡してきます。よって、**演算子を利用してメソッド定義をしていると、新しいRubyの処理系ではArgumentErrorが発生します。というわけで、こちらの方法も.initializeのオーバーライド時には利用できないということのようです。

Reply:
調査 & 実験ありがと。なるほど、これはやってみないとわからないやつですね。僕も知らなかったので振り回してしまってごめん。
(たぶん僕なら initialize の中で上書きしてしまうやりかたを取るので…)